packages/roosterjs-content-model/lib/publicApi/block/toggleMargins.ts

```diff
@@ -0,0 +1,29 @@
+import { getObjectKeys } from 'roosterjs-editor-dom';
+import { createParagraphDecorator } from '../../modelApi/creators/createParagraphDecorator';
+import { MarginFormat } from '../../publicTypes/format/formatParts/MarginFormat';
+import { IContentModelEditor } from '../../publicTypes/IContentModelEditor';
+import { formatParagraphWithContentModel } from '../utils/formatParagraphWithContentModel';
+
+/**
+ * Toggles the current block(s) margin properties.
+ * @param editor The editor to operate on
+ * @param marginFormat Object containing margin props to be changed.
+ * If a margin property is not in this object, it won't be changed.
+ */
+export default function toggleMargins(editor: IContentModelEditor, marginFormat: MarginFormat) {
+    formatParagraphWithContentModel(editor, 'toggleMargins', para => {
+        if (!para.decorator || para.decorator.tagName !== 'p') {
+            para.decorator = createParagraphDecorator('p');
+        }
+        getObjectKeys(marginFormat).forEach(key => {
+            if (
+                para.format[key] &&
+                (para.format[key] === marginFormat[key] || parseInt(para.format[key]!) > 0)
```

Comment:
why delete para.format[key] when format[key] == marginFormat[key]?

Reply:
To answer this and "I think name it "setMargins" is better", if we make it delete the margin format when format is already set, we can toggle the value, no need to get formatState to apply margin. Also removes a positive margin (coming from copy paste for example)

Reply:
Hmm, I think that make cause some confusion to user. If user doesn't want to change the margin, they need to be very careful to make sure not passing in a value that is the same with existing value. So user first need to get the existing value to check if it is the same with the one they want to check.

When a function is named "toggle", it means user doesn't need to know the original value, but just call it to turn on or off.

So still suggest:
1. name the function "setMargin" or "setParagraphMargin"
2. have two separate parameters for top and bottom
3. when pass in a valide value for top/bottom, apply it, no matter if it is the same with existing value
4. when pass in undefined, don't change
5. when pass in null or 0, delete the margin
so that is easier to understand and use.

Reply:
We have an API "toggleHeader" in original roosterjs code which is a bad example. In content model I have renamed it to be "setHeaderLevel"

Reply:
Got it, went ahead with all of your fixes
Had to add support for passing formatState in onClick callback for the spaceBeforeAfter buttons to remain simple